compiler/test/compilable/test23639.d

```diff
@@ -0,0 +1,12 @@
+// https://issues.dlang.org/show_bug.cgi?id=23639
+
+// REQUIRED_ARGS: -preview=nosharedaccess
+
+class T {}
+
+shared(T) allocClassMem()
+{
+    void *p;
+    // assume p is allocated here
+    return cast(shared(T))p;
```

Comment:
what is the `@safe`ty of this operation?

Reply:
Error: cast from `void*` to `shared(T)` not allowed in safe code

The expected @safety level - @system.

Reply:
and if `typeof(p) == T`?

Reply:
The modified code has nothing to do with the safety checks. Casts in from differently qualified types remain unsafe, that part is not modified.

Reply:
`-preview=nosharedaccess` is all about safety. even if the end result is that the code must be system. Perhaps I was thinking things wrong, but it seems to me that the only reason why you would want to do this with `-preview=nosharedaccess` is because you have a unique reference to an object you then want to share. Its fine if that needs to be trusted, then please disregard this request for changes, but please do document this discussion in the test case.

Reply:
Yes, a prime example of this is memory allocation. You allocate heap memory which is naturally shared among threads and you want to express that. What other way of casting the mallocd memory to shared is there?